src/dist/dist.rs

```diff
@@ -21,6 +21,25 @@ pub static DEFAULT_DIST_SERVER: &str = "https://static.rust-lang.org";
 // Deprecated
 pub static DEFAULT_DIST_ROOT: &str = "https://static.rust-lang.org/dist";
 
+// The channel patterns we support
+static TOOLCHAIN_CHANNELS: &[&str] = &[
+    "nightly",
+    "beta",
+    "stable",
+    // Allow from 1.0.0 through to 9.999.99
+    r"\d{1}\.\d{1,3}\.\d{1,2}",
+];
+
+lazy_static! {
+    static ref TOOLCHAIN_CHANNEL_PATTERN: String = format!(
+        r"^({})(?:-(\d{{4}}-\d{{2}}-\d{{2}}))?(?:-(.*))?$",
+        TOOLCHAIN_CHANNELS.join("|")
+    );
+    // Note this regex gives you a guaranteed match of the channel (1)
```

Comment:
Instead of writing this in a comment and referring to numeric capture group IDs, we could encode this in the type system, eg.

```rust
struct ParsedToolchainDesc {
    channel: String,
    date: Option<String>,
    target: Option<String>,
}

fn parse_toolchain_desc(input: &str) -> Result<ParsedToolchainDesc, Error> {
    ... local regex and extraction code ...
}
```

... and then this function could be used by both `FromStr` implementations. It would also make it easy to unit test the parsing function.

Reply:
You are quite right, and have seen through to my transparent attempt to leave nice refactoring jobs open for newcomers to the project.  If you'd rather I do this myself then I'm happy to, otherwise my intent was to file an issue describing something (perhaps in less concrete terms than your comment) to encourage someone to come and make a niceness happen as a way into the codebase.  In your view, should I make the change, or should I be filing that issue?  I'm happy either way.

Reply:
After two weeks I filed #2142